Add copy code button to code snippets

simonebortolin opened the pull request:
Hello everyone, this is my implementation for the copy button on the snippet (requested in #924)

The implementation is made 100% javascript as with or without a jekyll template modification you still have to execute some javascript code, and I consider it the best choice.

the button only appears if the mouse is over it, to allow the entire line to be read

the important CSS changes were made to make the copy button work even in the long code situation:


![image](https://user-images.githubusercontent.com/26844016/187731472-d4bf7828-2356-4d94-9c2d-9db863228f5f.png)

to avoid this:

![image](https://user-images.githubusercontent.com/26844016/183292313-d7f73d7d-58c0-4c7b-b5ba-e08bd285514b.png)

simonebortolin commented:
Yes of course the lint errors I wanted to solve them asap.

On Fri, 2 Sept 2022, 07:12 Matt Wang, ***@***.***> wrote:

> Hey @simonebortolin <https://github.com/simonebortolin>, thanks for
> submitting your PR! I'll give it a more in-depth look soon; in the
> meantime, could you resolve the lint errors?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/just-the-docs/just-the-docs/pull/945#issuecomment-1235072946>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGMZW4EODCQCRPAWELTGJXTV4GEDDANCNFSM6AAAAAAQBRPPSM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

simonebortolin commented:
Hello everyone, I should have corrected the style to make the lint happy (I hope, I have not yet seen the status of the automatic CI/CD controls)

I've also added an option to disable the copy button in the yml, I'm not sure if the logic is correct or if it's better != false for the default case management

simonebortolin commented:
I will add some thoughts/information

1) does it make sense to put a tooltip saying "copy made successfully", in my opinion no, it is very efficient the tick that appears and since the theme is very simple it does not make sense to complicate it even more with a tooltip, popper library and the like

2) the icons come from bootstrap icons, licensed by MIT, I don't think anything else needs to be added

3) why did I use the spread operator? since both the spread operator and the forEach on nodelist should be supported on all modern browsers with ES6.
No reason, it's personal taste, I find the syntax with the spread operator much nicer, of course I can remove it if you prefer without it.
I also know that 60% of programmers use it without the spread (https://twitter.com/chriscoyier/status/1060235806709735424)

4) is the onclick on copybutton ugly? de gustibus, it could be improved a lot by saving the timeout instance and checking if there is already one, and not allowing to make two copies at short distance, but does it make sense? I have no idea, you tell me, I prefer to keep it as simple as possible.

5) the value of the fourSeconds variable could well be moved to yml, it's something I've thought about a lot and I'm in favour of doing it, maybe I'll put it in the next commit of changes.

Translated with www.DeepL.com/Translator (free version)

simonebortolin commented:
I have just realised that I have broken the numbered lists for lines of code, I will correct this as soon as I can.

![image](https://user-images.githubusercontent.com/26844016/188424908-de3e45c3-8be6-426c-ba85-a03129cc8e2f.png)
![image](https://user-images.githubusercontent.com/26844016/188424985-0099bfea-0ae7-4746-830b-ae6cbe0ca48d.png)

mattxwang commented:
> Hello everyone, I should have corrected the style to make the lint happy (I hope, I have not yet seen the status of the automatic CI/CD controls)

FYI - you can run these locally as well with `npm test`, and you can autofix spacing, etc. with `npm test -- --autofix`!

> I have just realised that I have broken the numbered lists for lines of code, I will correct this as soon as I can.

No worries, take your time!

mattxwang commented:
Hey @simonebortolin, just wanted to check - were you hoping to merge this in for the upcoming release? If so, I think there's a merge conflict / CI issues, but after those are resolved happy to give a review!

simonebortolin commented:
Hi, after September 21 I will fix everything.

Now I really don't think that feature will come out in v4.0.0.



On Thu, 15 Sept 2022, 02:16 Matt Wang, ***@***.***> wrote:

> Hey @simonebortolin <https://github.com/simonebortolin>, just wanted to
> check - were you hoping to merge this in for the upcoming release? If so, I
> think there's a merge conflict / CI issues, but after those are resolved
> happy to give a review!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/just-the-docs/just-the-docs/pull/945#issuecomment-1247424531>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGMZW4DXFYODKLAA2DQTPXDV6JTHHANCNFSM6AAAAAAQBRPPSM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mattxwang commented:
No worries, sounds good!

simonebortolin commented:
I've already sketched out the fix anyway, it just needs to be finished and
committed

Simone


On Thu, 15 Sept 2022, 20:44 Matt Wang, ***@***.***> wrote:

> No worries, sounds good!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/just-the-docs/just-the-docs/pull/945#issuecomment-1248470684>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGMZW4HLE5HXCR2SSFINAWLV6NU7HANCNFSM6AAAAAAQBRPPSM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

simonebortolin commented:
Hello @mattxwang everything should now be complete

I should have synchronised the repo, I rebased a few commits because there were so many to manage in one PR 

And I launched `npm test`, fixed everything I modified, but I get errors on code I haven't touched.

simonebortolin commented:
@mattxwang can you try to launch the workflow? these days I am free and will fix any problems

simonebortolin commented:
@mattxwang can you try again

mattxwang commented:
Will give a review in a bit - quite swamped with non-jtd obligations unfortunately!

mattxwang reviewed (changes requested):
Hey @simonebortolin, a couple of broad stroke notes:

- if you're adding a configuration option, we should add some documentation explaining how users could use it. Can you add some writing to the configuration page to explain how to use this feature?
- what's the driving use-case for customizing the timeout time? I imagine it would be relatively standardized?
> the icons come from bootstrap icons, licensed by MIT, I don't think anything else needs to be added
- this is not true - if the icons are MIT licensed, we're obligated to redistribute a copy of the license (and, additionally, *we should credit them*)
- there are some areas of the site where the new styling you've added changes the spacing in a code block. For example, [on the code page on `main`](https://just-the-docs.github.io/just-the-docs/docs/ui-components/code/):
<img width="740" alt="Screen Shot 2022-09-25 at 6 11 03 PM" src="https://user-images.githubusercontent.com/14893287/192175382-a3d0af50-a737-4a13-b688-4c34f5607cda.png">

[The same page on the deploy preview:](https://deploy-preview-945--just-the-docs.netlify.app/docs/ui-components/code/)
<img width="736" alt="Screen Shot 2022-09-25 at 6 11 14 PM" src="https://user-images.githubusercontent.com/14893287/192175446-a7a7a1dd-6993-454e-a722-6a5e890a9888.png">

Note the shifted backticks/js!

Appreciate your work on this PR. I think with a bit of tidying, we'll be able to merge in no time!

simonebortolin commented:
@mattxwang 

> * if you're adding a configuration option, we should add some documentation explaining how users could use it. Can you add some writing to the configuration page to explain how to use this feature?

done

> * what's the driving use-case for customizing the timeout time? I imagine it would be relatively standardized?

none, but one more option is always better than less

> > the icons come from bootstrap icons, licensed by MIT, I don't think anything else needs to be added
> * this is not true - if the icons are MIT licensed, we're obligated to redistribute a copy of the license (and, additionally, _we should credit them_)

yes, but JTD is also via MIT licence, so the licence copy is already there.  I don't think anything more needs to be said, maybe I'm wrong.

> * there are some areas of the site where the new styling you've added changes the spacing in a code block. For example, [on the code page on `main`](https://just-the-docs.github.io/just-the-docs/docs/ui-components/code/):

done

mattxwang commented on docs/ui-components/code.md:
```diff
@@ -143,3 +143,15 @@ graph TD;
+
+## Copy button
+
+Copy button can enable or disable via `copy_button_on_code` entry in `_config.yml`, as well as set the action feedback timeout, via `copy_button_on_code_time` entry in `_config.yml`.
```
```suggestion
The copy button for code blocks can be enabled or disabled via the `copy_button_on_code` key in `_config.yml`. By default, the value of this key is `false`; users need to opt-in.
```

mattxwang commented on docs/ui-components/code.md:
````diff
@@ -143,3 +143,15 @@ graph TD;
+copy_button_on_code_time: 4000
+```
+
+Generally the timeout time should be between 3 and 5 seconds, if not specified the default value is 4 seconds. In general, there is no real reason to change this value.
````
```suggestion
Note that this feature requires JavaScript; if JavaScript is disabled in the browser, this feature will not work.
```

mattxwang reviewed (changes requested):
Sorry, have been in and out - personal stuff.

> none, but one more option is always better than less

Hm, I slightly disagree, since this is extra stuff we need to document and maintain. If there's no motivating use-case, can we not include this for now? It's much easier for us to add features later on, once people request it.

> yes, but JTD is also via MIT licence, so the licence copy is already there. I don't think anything more needs to be said, maybe I'm wrong.

Ah, not exactly. We need to include a different license copy for each different project we use; this is because each library has its own copyright notice attached to the MIT license. For example, this repo's reads:

> Copyright (c) 2016 Patrick Marsceill

Which is the original creator of the theme. In contrast, [the Bootstrap Icon MIT License](https://github.com/twbs/icons/blob/main/LICENSE.md) reads

> Copyright (c) 2019-2021 The Bootstrap Authors

Light-touch changes left, but if we address those two points I'd be happy to merge this in! Apologies for how long this has taken, usually my iteration loop is tighter.

olgarithms commented on assets/js/just-the-docs.js:
```diff
@@ -479,6 +479,49 @@ jtd.onReady(function(){
+
+        copyButton.innerHTML = svgCopied;
+
+        var fourSeconds = {{ site.copy_button_on_code_time | default: 4000 }};
```
This variable name can be misleading if a user sets it to anything different from 4s? Maybe `copyButtonTimeout` or `timeoutSetting` would be better?

olgarithms commented on _config.yml:
```diff
@@ -51,6 +51,10 @@ search:
 
+# For copy button on code
+copy_button_on_code: true
+copy_button_on_code_time: 4000
```
Renaming this variable `copy_button_on_code_timeout` could be more expressive and lead to fewer questions?

olgarithms reviewed (commented):
Thanks @simonebortolin for implementing this feature. As a user of the theme I am very interested, so I took a look and left a couple of comments in regards to naming:

simonebortolin commented:
> Thanks @simonebortolin for implementing this feature. As a user of the theme I am very interested, so I took a look and left a couple of comments in regards to naming:

yes you are right, but @mattxwang told me to remove it

@mattxwang sorry I've been full too, next week I'll make the changes

mattxwang commented:
No worries, take your time @simonebortolin!

simonebortolin commented:
@mattxwang OK I'm there
- I added the image licence following the style of https://github.com/just-the-docs/just-the-docs/blob/main/_includes/icons/external_link.html
- I have apply your and @olgarithms suggestion
- I rebased everything with the last commit in main branch

max06 commented on _config.yml:
```diff
@@ -73,6 +73,9 @@ search:
   button: false
 
+# For copy button on code
+copy_button_on_code: true
```
Little nitpicking here: I'd prefer `enable_copy_code_button` - `enable` indicates a boolean (true/default-false), `copy_code` is an action. More speaking to me.

simonebortolin replied:
you're right

mattxwang commented on assets/js/just-the-docs.js:
```diff
@@ -479,6 +479,49 @@ jtd.onReady(function(){
+
 })(window.jtd = window.jtd || {});
 
+
+
```
nit: this is unnecessary whitespace, right?
```suggestion
```

mattxwang commented on _sass/code.scss:
```diff
@@ -48,17 +50,61 @@ a:visited code {
+  position: relative;
+  padding: 0;
+
+  > button {
```
Could you add a CSS comment explaining what's going on here? I don't think it's immediately obvious to a developer. One or two sentences suffices!

(it's also not immediately clear to me why all of these properties are needed. have you done any pruning on them?)

simonebortolin replied:
I investigate

simonebortolin replied:
the only thing I can remove is the height... which solves the problem on mobile

mattxwang commented on _sass/code.scss:
```diff
@@ -48,17 +50,61 @@ a:visited code {
+
+  &:hover {
+    > button {
+      opacity: 1;
```
nit: makes it more obvious what this feature does

```suggestion
      cursor: copy;
      opacity: 1;
```

mattxwang commented on _sass/code.scss:
```diff
@@ -48,17 +50,61 @@ a:visited code {
+    right: 0;
+    border: $sp-3 solid $code-background-color;
+    background-color: $code-background-color;
+    outline: none;
```
```suggestion
    outline: none;
```

Is there a reason why this needs to be here? If we remove this, then the button has a built-in browser focus state; see this in chrome:

<img width="768" alt="Screenshot 2022-12-26 at 2 31 15 PM" src="https://user-images.githubusercontent.com/14893287/209586595-7e05730c-e397-4c72-b92a-07c8912e69e2.png">

mattxwang reviewed (commented):
Thanks again @simonebortolin for taking this up, and don't apologize at all for taking a while - this is free open-source, and you're volunteering your time!

Overall I think this is pretty good to merge; I've left some minor nits in code review. I've also noticed on mobile that the clipboard gets clipped; this is me on Safari on iOS:

![321927058_1229400581291767_8585685839582402084_n](https://user-images.githubusercontent.com/14893287/209586220-9ba1100f-4b96-4e5a-ad53-a8d48e021c16.jpg)

To be frank, I'm actually fine with merging this as-is with the mobile bug (and fixing it in a follow-up), since I think it's not breaking the feature + it's uncommon for users to copy code from a mobile device. If you can address the suggested comments, we can merge this in for the last RC of this calendar year!

mattxwang reviewed (approved):
Leaving a green check since this is technically in a mergeable state, but @simonebortolin it would be great if you could also prune the CSS properties and/or add some docs.

I'll merge this in before releasing our RC & resolve the merge conflict with #1058 (should be trivial).

simonebortolin commented:
@mattxwang OK then 
- I have put some documentation, if you do not find it sufficient I will fix all comments according to the new logic
- It is very difficult to reduce the css code, it is all written in the first paragraph because a massive edit was necessary, as the way the css code was structured before made it impossible to add the button.
- for #1058 do I have to do it?

simonebortolin commented:
@mattxwang I have added comments explaining the logic of how code.scss works

bitterly unable to reduce the number, I gave the best of my knowledge to get this PR right

mattxwang commented:
No worries, thanks for taking this on!

> do I have to do it?

Nope, I'll handle it since I'm authoring the other PR!

mattxwang commented:
Thanks for taking this on! Much appreciated @simonebortolin, and congrats on your first contribution!

simonebortolin commented:
>No worries, thanks for taking this on!

it was beautiful!

> 

Thanks for taking this on! Much appreciated @simonebortolin, and congrats on your first contribution!

Thanks! come on that the second PR is already in the working

simonebortolin commented on assets/js/just-the-docs.js:
```diff
@@ -479,6 +479,47 @@ jtd.onReady(function(){
+
+    copyButton.addEventListener('click', function () {
+      if(timeout === null) {
+        var code = codeBlock.querySelector('code').innerText.trim();
```
@pdmosses i think the problem is here
